s3/pkg/meta/redis/redis.go

```diff
@@ -0,0 +1,385 @@
+	c, err := GetClient()
+	if err != nil {
+		helper.Logger.Printf(2, "failed to get redis client, err: %v", err)
+		return err
+	}
+	encodedValue, err := helper.MsgPackMarshal(value)
+	if err != nil {
+		helper.Logger.Printf(2, "failed to make pack for(%s, %s, %v), err: %v", prefix, key, value, err)
+		return err
+	}
+
+	// Use table.String() + hashkey as Redis key. Set expire time to 30s.
+	r, err := c.Set(table.String()+prefix+helper.EscapeColon(key), encodedValue, 30000)
+	if err != nil {
+		helper.Logger.Printf(2, "failed to call redis Set(%s, %v), err: %v", table.String()+prefix+helper.EscapeColon(key), string(encodedValue), err)
+		return err
+	}
+	helper.Logger.Printf(20, "Cmd: %s. Key: %s. Value: %s. Reply: %s.", "SET", table.String()+key, string(encodedValue), r)
+	return nil
+}
+
+func Get(table RedisDatabase, prefix, key string) (value interface{}, err error) {
+	var encodedValue []byte
+	c, err := GetClient()
+	if err != nil {
+		helper.Logger.Printf(2, "failed to get redis client, err: %v", err)
+		return nil, err
+	}
+
+	// Use table.String() + hashkey as Redis key
+	encodedValue, err = c.Get(table.String() + prefix + helper.EscapeColon(key))
+	if err != nil {
+		helper.Logger.Printf(2, "failed to call redis Get(%s), err: %v", table.String()+prefix+helper.EscapeColon(key), err)
+		return nil, err
+	}
+	if len(encodedValue) == 0 {
+		return nil, nil
+	}
+	err = helper.MsgPackUnMarshal(encodedValue, value)
+	return value, err
+}
+
+// don't use the escapecolon in keys command.
+func Keys(table RedisDatabase, pattern string) ([]string, error) {
+	var keys []string
+	query := table.String() + pattern
+	c, err := GetClient()
+	if err != nil {
+		helper.Logger.Printf(2, "failed to get redis client, err: %v", err)
+		return nil, err
+	}
+
+	keys, err = c.Keys(query)
+	if err != nil {
+		helper.Logger.Printf(2, "failed to call redis Keys(%s), err: %s", query, err)
+		return nil, err
+	}
+
+	return keys, nil
+
+}
+
+func MGet(table RedisDatabase, prefix string, keys []string) ([]interface{}, error) {
+	var results []interface{}
+	var queryKeys []string
+	for _, key := range keys {
+		queryKeys = append(queryKeys, table.String()+prefix+helper.EscapeColon(key))
+	}
+	c, err := GetClient()
+	if err != nil {
+		helper.Logger.Printf(2, "failed to get redis client, err: %v", err)
+		return nil, err
+	}
+
+	results, err = c.MGet(queryKeys)
+	if err != nil {
+		helper.Logger.Printf(2, "failed to call redis MGet(%v), err: %v", queryKeys, err)
+		return nil, err
+	}
+	return results, nil
+}
+
+func MSet(table RedisDatabase, prefix string, pairs map[string]interface{}) (string, error) {
+	var result string
+	tmpPairs := make(map[interface{}]interface{})
+	for k, v := range pairs {
+		tmpPairs[table.String()+prefix+helper.EscapeColon(k)] = v
+	}
+
+	c, err := GetClient()
+	if err != nil {
+		helper.Logger.Printf(2, "failed to get redis client, err: %v", err)
+		return "", err
+	}
+
+	result, err = c.MSet(tmpPairs)
```

Comment:
how does this work if, from the many keys, some are set and some cannot be set ?